### PR TITLE
Add user facing documentation for CSAN

### PR DESCRIPTION
**Pull request by sypneiwski**
This adds a user facing tutorial for the CSAN tool. The documentation preview should be available [here](https://docs-preview.pytorch.org/84689/index.html) once the GitHub job completes on this PR.

### CODE REVIEW

File: docs/source/cuda._sanitizer.rst [hunk 1]
```diff
@@ -0,0 +1,95 @@
+.. currentmodule:: torch.cuda._sanitizer
+
+torch.cuda._sanitizer
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:3]: The title could be in plain English, right? "CUDA stream sanitizer"?

File: docs/source/cuda._sanitizer.rst [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+=====================
+
+Overview
+--------
+
+.. automodule:: torch.cuda._sanitizer
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:9]: What exactly does this show? Is it an overview of the classes/functions/members of the module? If so, then this may be a bit "harsh" introduction to put as the very first thing the reader sees when trying to understand what CSAN even is.

On the other hand, if this is just the module's docstring, without anything else, then it may be fine. However I haven't read that docstring yet...

File: docs/source/cuda._sanitizer.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+    import torch
+
+    a = torch.rand(4, 2, device="cuda")
+
+    with torch.cuda.stream(torch.cuda.Stream()):
+        torch.mul(a, 5, out=a)
+
+The ``a`` tensor is initialized on the default stream and, without any synchronization
+methods, modified on a new stream. When this script is run on the commandline with:
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:27]: Explain why this is a bug. Say it clearly: the two kernels will run concurrently on the same data, thus the second kernel might read some uninitialized data before the first kernel was able to write it, or the first kernel might even overwrite some data of the second kernel.

File: docs/source/cuda._sanitizer.rst [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+the following output is printed by CSAN:
+
+::
+
+    ============================
+    CSAN detected a possible data race on tensor with data pointer 139719969079296
+    Access by stream 94646435460352 during kernel:
+    aten::mul.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
+    writing to argument: self, out, output
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:40]: [nit] I'm just seeing it now: it says "writing to argument output", but there is no argument called output. I guess it means that the problematic tensor is the output of the operator. Can it be made clearer? Something like "writing to arguments self and out, and to the output"?

File: docs/source/cuda._sanitizer.rst [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    With stack trace:
+        File "pytorch/torch/cuda/_sanitizer.py", line 364, in _handle_kernel_launch
+            stack_trace = traceback.StackSummary.extract(
+        ...
+        File "example_error.py", line 6, in <module>
+            torch.mul(a, 5, out=a)
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:46]: I thought in Python the innermost frame is printed last, whereas here it comes first. Also by default traceback indents lines by two spaces, not four. Is this literally a verbatim output of CSAN, or did you build it by hand?

File: docs/source/cuda.rst [hunk 6]
```diff
@@ -134,3 +134,14 @@ Jiterator (beta)
 
     jiterator._create_jit_fn
     jiterator._create_multi_output_jit_fn
+
+Sanitizer (prototype)
```
Comment (lw) [inline docs/source/cuda.rst:138]: Maybe "Stream sanitizer", to clarify it's not a silver bullet that sanitizes every possible aspect of CUDA?

File: docs/source/cuda._sanitizer.rst [hunk 7]
```diff
@@ -0,0 +1,95 @@
+.. currentmodule:: torch.cuda._sanitizer
+
+torch.cuda._sanitizer
+=====================
+
```
Comment (lw) [inline docs/source/cuda._sanitizer.rst:5]: I think it would be good to add a box/warning at the very top that this is a prototype feature, with some explanation of what it means in terms of support, of stability, ...

It would also be nice, later on, to have a kind of list of known feature gaps, i.e., things that CSAN doesn't yet support.

**Review by lw — APPROVED** [review]

File: docs/source/cuda._sanitizer.rst (hunk 2 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:9]: This is the module's docstring, without anything else. I've made sure it serves as a good introduction to the page.

File: docs/source/cuda._sanitizer.rst (hunk 5 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:46]: This is a _slightly modified_ version of the verbatim output of CSAN, with the frames in the middle abbreviated for clarity. The indentation is a mistake on my part, I will fix it.

The order of the frames is just as CSAN prints it, which is how [`StackSummary`](https://docs.python.org/3/library/traceback.html#traceback.StackSummary) formats it.

File: docs/source/cuda.rst (hunk 6 above)
Comment (sypneiwski) [reply docs/source/cuda.rst:138]: You're right, I've changed it.

File: docs/source/cuda._sanitizer.rst (hunk 1 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:3]: I thought the convention was to use the module name; but it seems the two other pages I saw that did this were the only ones. I will change it.

File: docs/source/cuda._sanitizer.rst (hunk 3 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:27]: Done.

File: docs/source/cuda._sanitizer.rst (hunk 4 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:40]: I have a future rework in mind that will address this issue, but I will add this as a separate PR.

File: docs/source/cuda._sanitizer.rst (hunk 5 above)
Comment (sypneiwski) [reply docs/source/cuda._sanitizer.rst:46]: I've reversed the stored `StackSummary` objects.
